distribution/lib/Standard/Base/0.0.0-dev/src/System/File/Existing_File_Behavior.enso

```diff
@@ -0,0 +1,89 @@
+       later.
+
+       The `action` may not be run at all in case the `Error` behavior is
+       selected.
+    write : File -> (Output_Stream -> Nothing) -> Nothing ! File_Not_Found | Io_Error | File_Already_Exists_Error
+    write file action =
+        case this of
+            Overwrite -> file.with_output_stream [Option.Write, Option.Create, Option.Truncate_Existing] action
+            Append -> file.with_output_stream [Option.Write, Option.Create, Option.Append] action
+            Error -> file.with_output_stream [Option.Write, Option.Create_New] action
+            Backup -> Panic.recover [Io_Error, File_Not_Found] <|
+                handle_existing_file _ =
+                    here.write_file_backing_up_old_one file action
+                ## We first attempt to write the file to the original
+                   destination, but if that files due to the file already
+                   existing, we will run the alternative algorithm which uses a
+                   temporary file and creates a backup.
+                Panic.catch File_Already_Exists_Error handler=handle_existing_file <|
+                    Panic.rethrow <| file.with_output_stream [Option.Write, Option.Create_New] action
+
+## PRIVATE
+write_file_backing_up_old_one : File -> (Output_Stream -> Nothing) -> Nothing ! File_Not_Found | Io_Error | File_Already_Exists_Error
+write_file_backing_up_old_one file action = Panic.recover [Io_Error, File_Not_Found] <|
+    parent = file.parent
+    bak_file = parent / file.name+".bak"
+    go i =
+        new_name = file.name + ".new" + if i == 0 then "" else "." + i.to_text
+        new_file = parent / new_name
+        handle_existing_file _ = go i+1
+        handle_write_failure panic =
+            ## Since we were already inside of the write operation,
+               the file must have been created, but since we failed, we need to clean it up.
+            new_file.delete
+            Panic.throw panic.payload.cause
+        Panic.catch File_Already_Exists_Error handler=handle_existing_file <|
+            Panic.catch Internal_Write_Operation_Failed handler=handle_write_failure <|
+                Panic.rethrow <|
+                    new_file.with_output_stream [Option.Write, Option.Create_New] output_stream->
+                        Panic.catch Any (action output_stream) caught_panic->
+                            Panic.throw (Internal_Write_Operation_Failed caught_panic)
+            ## We ignore the file not found error, because it means that there
+               is no file to back-up. This may also be caused by someone
+               removing the original file during the time when we have been
+               writing the new one to the temporary location. There is nothing
+               to back-up anymore, but this is not a failure, so it can be
+               safely ignored.
+            Panic.catch File_Not_Found handler=(_->Nothing) <|
+                Panic.rethrow <| file.move_to bak_file
```

Comment:
That would potentially overwrite an already existing backup file?

Reply:
Yes, that is the design, as [documented in the doc comment for `type Backup`](https://github.com/enso-org/enso/pull/3518/files#diff-d2eeefd9bbce42c138e36fdf7782bb7e956d910c39f6f2c40e055e4375ddb869R19) which comes from the design doc.